src/metabase/task/persist_refresh.clj

```diff
@@ -0,0 +1,369 @@
+            [clojurewerkz.quartzite.conversion :as qc]
+            [clojurewerkz.quartzite.jobs :as jobs]
+            [clojurewerkz.quartzite.schedule.cron :as cron]
+            [clojurewerkz.quartzite.triggers :as triggers]
+            [java-time :as t]
+            [metabase.driver.ddl.interface :as ddl.i]
+            [metabase.models.card :refer [Card]]
+            [metabase.models.database :refer [Database]]
+            [metabase.models.persisted-info :as persisted-info :refer [PersistedInfo]]
+            [metabase.models.task-history :refer [TaskHistory]]
+            [metabase.public-settings :as public-settings]
+            [metabase.task :as task]
+            [metabase.util :as u]
+            [metabase.util.i18n :refer [trs]]
+            [potemkin.types :as p]
+            [toucan.db :as db])
+  (:import [org.quartz ObjectAlreadyExistsException Trigger]))
+
+(defn- job-context->job-type
+  [job-context]
+  (select-keys (qc/from-job-data job-context) ["db-id" "persisted-id" "type"]))
+
+(def refreshable-states
+  "States of `persisted_info` records which can be refreshed."
+  #{"persisted" "error"})
+
+(p/defprotocol+ Refresher
+  "This protocol is just a wrapper of the ddl.interface multimethods to ease for testing. Rather than defing some
+   multimethods on fake engine types, just work against this, and it will dispatch to the ddl.interface normally, or
+   allow for easy to control custom behavior in tests."
+  (refresh! [this database definition dataset-query])
+  (unpersist! [this database persisted-info]))
+
+(def dispatching-refresher
+  "Refresher implementation that dispatches to the multimethods in [[metabase.driver.ddl.interface]]."
+  (reify Refresher
+    (refresh! [_ database definition dataset-query]
+      (binding [persisted-info/*allow-persisted-substitution* false]
+        (ddl.i/refresh! (:engine database) database definition dataset-query)))
+    (unpersist! [_ database persisted-info]
+     (ddl.i/unpersist! (:engine database) database persisted-info))))
+
+(defn- refresh-with-state! [refresher database persisted-info]
+  (when-not (= "deletable" (:state persisted-info))
+    (let [card (Card (:card_id persisted-info))
+          definition (persisted-info/metadata->definition (:result_metadata card)
+                                                          (:table_name persisted-info))
+          _ (db/update! PersistedInfo (u/the-id persisted-info)
+                        :definition definition,
+                        :query_hash (persisted-info/query-hash (:dataset_query card))
+                        :active false,
+                        :refresh_begin :%now,
+                        :refresh_end nil,
+                        :state "refreshing"
+                        :state_change_at :%now)
+          {:keys [state] :as results} (refresh! refresher database definition (:dataset_query card))]
+      (if (= state :success)
+        (db/update! PersistedInfo (u/the-id persisted-info)
```

Comment:
Since `db/update!` can take `model id m` maybe it makes sense to push the `if` inside a single call to `db/update!`